### ISSUE TITLE
UI and agent driven content negotiation for language selection not applied upon filtering facets labels

### ISSUE THREAD

**Issue by sgrellet** [UKGovLD/registry-core#162]
Bonjour eveyrone, 

We are currently facing a translation issue with LDRegistry UI: agent driven content negotiation for language selection is not applied upon filtering facets labels. 
@https://github.com/cfollenf : please complete when needed

We have a list of “categories” where each instance has its rdfs:label and skos:prefLabel values specified in two languages, namely English and French. These categories are supposed to be used as facets, and their instances look like those:

~~~~

<https://data.geoscience.fr/ncl/structure/Category/1>
  rdf:type skos:Concept ;
  dct:description "Catégorie permettant d'identifier les registres et/ou entités utilisés pour l'eau et la biodiversité"@fr ;
  reg:notation "1" ;
  rdfs:label "eau et biodiversité"@fr ;
  rdfs:label "water and biodiversity"@en ;
  skos:inScheme <https://data.geoscience.fr/ncl/Category> ;
  skos:notation "WaterBiodiversity" ;
  skos:prefLabel "eau et biodiversité"@fr ;
  skos:prefLabel "water and biodiversity"@en ;
  adms:status <http://inspire.ec.europa.eu/registry/status/valid> .
<https://data.geoscience.fr/ncl/structure/Category/4>
  rdf:type skos:Concept ;
  dct:description "Catégorie permettant d'identifier les registres et/ou entités utilisés pour les Eaux Soutteraines"@fr ;
  reg:notation "4" ;
  rdfs:label "eaux souterraines"@fr ;
  rdfs:label "groundwater"@en ;
  skos:inScheme <https://data.geoscience.fr/ncl/Category> ;
  skos:notation "Groundwater" ;
  skos:prefLabel "eaux souterraines"@fr ;
  skos:prefLabel "groundwater"@en ;
  adms:status <http://inspire.ec.europa.eu/registry/status/valid .
<https://data.geoscience.fr/ncl/structure/Category/6>
  rdf:type skos:Concept ;
  dct:description "Catégorie permettant d'identifier les registres et/ou entités utilisés dans le cadre de la description de diagraphies"@fr ;
  reg:notation "6" ;
  rdfs:label "diagraphie"@fr ;
  rdfs:label "diagraphy"@en ;
  skos:inScheme <https://data.geoscience.fr/ncl/Category> ;
  skos:notation "Diagraphy" ;
  skos:prefLabel "diagraphie"@fr ;
  skos:prefLabel "diagraphy"@en ;
  adms:status <http://inspire.ec.europa.eu/registry/status/valid> ;

~~~~

When querying the LDRegistry web application, this UI is returned:
![image](https://user-images.githubusercontent.com/16756304/176933711-463c5aeb-798e-46bf-a1bd-52fdb1387628.png)

As you can see, categories listed in the left sidebar are randomly labelled with either French or English values, while both are always present in the data, no matter the fact English is selected for language content negotiation and correctly applied to registries in the main view.
We think the SPARQL query generated to retrieve these labels could be the problem source, as using the SPARQL endpoint to query the categories data with

~~~
SELECT * WHERE {
       VALUES ?cat { <https://data.geoscience.fr/ncl/structure/Category/3> 
                  <https://data.geoscience.fr/ncl/structure/Category/6>
                  <https://data.geoscience.fr/ncl/structure/Category/1>
                  <https://data.geoscience.fr/ncl/structure/Category/13>
                  <https://data.geoscience.fr/ncl/structure/Category/4>
                  <https://data.geoscience.fr/ncl/structure/Category/2>
                  <https://data.geoscience.fr/ncl/structure/Category/14>
                  <https://data.geoscience.fr/ncl/structure/Category/5>
                  <https://data.geoscience.fr/ncl/structure/Category/9>
                  <https://data.geoscience.fr/ncl/structure/Category/7>
                }
       ?cat rdfs:label ?y .
}
~~~

returns the result set below:
![image](https://user-images.githubusercontent.com/16756304/176933822-43867930-e605-4d31-804c-2186feba5074.png)

For each category subject, its first label object corresponds to the one displayed in the UI, thus it looks like the application generates a label retrieval query that merely lists all values without any language filtering and picks the first triples in the result set, applying an ascending ordering to ?cat values and the default ordering returned by the basestore to ?y values.

Is there a language content negotiation aspect missing on the facet labels query generator?

**Comment by der** [comment]
You are correct. The facet system doesn't support language negotiation, if faced with multiple labels it just picks the first on the list which will be effectively random. That seems to have been missed in previous rounds of improving language support. Though we know that some users replace the landing page and prefer not to use the facet system.

Looking at the code the lexical presentation of facets and their values are obtained using `RDFUtil.getLabel` in plain form rather than passing in a language to match on. 

So the facet system would need extending to pass in a preferred language and pass that to the `getLabel` calls. Not sure, without more investigation, whether than could be done under the hood using session attributes for the language or whether would also need some matching changes to `registry-config-base`.

**Comment by simonoakesepimorphics** [comment]
I have released a snapshot version of registry-core 2.3.9 with this feature. You can find the WAR file [here](https://s3-eu-west-1.amazonaws.com/ukgovld/snapshot/com/github/ukgovld/registry-core/2.3.9-SNAPSHOT/registry-core-2.3.9-20220720.155944-1.war). In order for the language negotiation to apply to facets, you must have the change from [this PR](https://github.com/UKGovLD/registry-config-base/pull/31) in your Velocity templates (currently available in the master branch of registry-config-base).

**Comment by simonoakesepimorphics** [comment]
Version 2.3.9 has been released with this feature.
